### DISCUSSION TITLE
Multiple patches in a single transaction?

### DISCUSSION THREAD

**Discussion by cadilhac** [ravendb/ravendb#19329] [Q&A]
I already have this operation:

```C#
var timelineId = "timelines/" + viewModel.TimelineId;
var query = "from Images as image" +
	$" where image.TimelineId == \"{timelineId}\"" +
	" update { image.Ghost = true; }";

var operation = RavenSession.Advanced.DocumentStore.Operations
	.Send(new PatchByQueryOperation(query));

operation.WaitForCompletion();
```

I would like to know if it's possible to have a second operation being performed in the same transaction. This operation has to patch 2 properties of a single document. I can't find the syntax to do that.

**Comment by ayende** [comment]
1/ Note that `PatchByQueryOperation` is not a global transaction. That is - it works in batches, and each batch is transactional, but if you have 100'000 items, we'll not modify them all in one shot.

2/ You are using string concat for this, which is a _bad_ idea. 

3/ Why aren't you modifying two properties?


```
var timelineId = "timelines/" + viewModel.TimelineId;
var query = @"
from Images as image
where image.TimelineId == $id
update { 
    image.Ghost = true; 
    image.Color = 'gray'; 
}
";

var operation = RavenSession.Advanced.DocumentStore.Operations
	.Send(new PatchByQueryOperation(new IndexQuery { Query =query , QueryParameters = {["id"] = timelineId }});;

operation.WaitForCompletion();
```

**Reply by cadilhac** [reply]
2. What do you mean?
3. The second operation updates 2 props in another document completely unrelated to the first op.

**Reply by ayende** [reply]
2/ This isn't recommended: ` $" where image.TimelineId == \"{timelineId}\"" +`
3/ Can you show what you _want_ to do completely?

**Reply by cadilhac** [reply]
2. Oh, this is unrelated to RavenDb then... Can you educate me on this if time permits? Were you thinking about the fact that several string instances are being created for this code and that it's not very performant?
3. Right now, I'm doing this:

```C#
var timelineId = "timelines/" + viewModel.TimelineId;
var query = "from Images as image" +
	$" where image.TimelineId == \"{timelineId}\"" +
	" update { image.Ghost = true; }";

var operation = RavenSession.Advanced.DocumentStore.Operations
	.Send(new PatchByQueryOperation(query));

operation.WaitForCompletion();

// Patch timeline
RavenSession.Advanced.Patch<Timeline, bool?>(timelineId, x => x.IsEncrypted, null);
RavenSession.Advanced.Patch<Timeline, DateTime>(timelineId, x => x.LastModified, DateTime.UtcNow);
await RavenSession.SaveChangesAsync();
```

As you can see, this is an op, and then a session save.

**Reply by ayende** [reply]
2/ Think - SQL Injection.

As for your PATCH, you can do:

```
var timelineId = "timelines/" + viewModel.TimelineId;
var query = @"
from Images as image
where image.TimelineId == $id
update { 
    let timeline = load($id);
    timeline.IsEncrypted = null;
    timeline.LastModified = new Date();
    put($id, timeline);

    image.Ghost = true; 
    image.Color = 'gray'; 
}
";

var operation = RavenSession.Advanced.DocumentStore.Operations
	.Send(new PatchByQueryOperation(new IndexQuery { Query =query , QueryParameters = {["id"] = timelineId }});;

operation.WaitForCompletion();
```


And that would modify it as part of the same operation

**Reply by cadilhac** [reply]
Thank you
